music/src/piano_genie/model.ts

```diff
@@ -0,0 +1,295 @@
+    state.h.push(tf.zeros([1, RNN_NUNITS], 'float32'));
+  }
+  return state;
+}
+
+/**
+ * Frees LSTM state from GPU memory.
+ *
+ * @param state: The LSTM state to free.
+ */
+function disposeState(state: LSTMState) {
+  for (let i = 0; i < state.c.length; ++i) {
+    state.c[i].dispose();
+    state.h[i].dispose();
+  }
+}
+
+/**
+ * Piano Genie model class.
+ */
+class PianoGenie {
+  private checkpointURL: string;
+  private initialized: boolean;
+
+  // Model state
+  private modelVars: { [varName: string]: tf.Tensor };
+  private decLSTMCells: tf.LSTMCellFunc[];
+  private decForgetBias: tf.Scalar;
+  private state: LSTMState;
+
+  /**
+   * Piano Genie constructor.
+   *
+   * @param checkpointURL Path to the checkpoint directory.
+   */
+  constructor(checkpointURL: string) {
+    this.checkpointURL = checkpointURL;
+    this.initialized = false;
+  }
+
+  /**
+   * Resets Piano Genie LSTM state.
+   */
+  resetState() {
+    if (this.state !== undefined) {
+      disposeState(this.state);
+    }
+    this.state = createZeroState();
+  }
+
+  /**
+   * Returns whether or not the model has been initialized.
+   */
+  isInitialized() {
+    return this.initialized;
+  }
+
+  /**
+   * Load model weights.
+   *
+   * @param staticVars Optional pre-loaded weights for testing.
+   */
+  async initialize(staticVars?: tf.NamedTensorMap) {
+    if (this.initialized) {
+      this.dispose();
+    }
+
+    if (this.checkpointURL === undefined && staticVars === undefined) {
+      throw new Error('Need to specify either URI or static variables');
+    }
+
+    if (staticVars === undefined) {
+      const vars = await fetch(`${this.checkpointURL}/weights_manifest.json`)
+        .then((response) => response.json())
+        .then(
+          (manifest: tf.io.WeightsManifestConfig) =>
+            tf.io.loadWeights(manifest, this.checkpointURL));
+      this.modelVars = vars;
+    } else {
+      this.modelVars = staticVars;
+    }
+
+    this.decLSTMCells = [];
+    this.decForgetBias = tf.scalar(1, 'float32');
+    for (let i = 0; i < RNN_NLAYERS; ++i) {
+      const cellPrefix =
+        `phero_model/decoder/rnn/rnn/multi_rnn_cell/cell_${i}/lstm_cell/`;
+
+      this.decLSTMCells.push(
+        (data: tf.Tensor2D, c: tf.Tensor2D, h: tf.Tensor2D) =>
+          tf.basicLSTMCell(
+            this.decForgetBias,
+            this.modelVars[cellPrefix + 'kernel'] as tf.Tensor2D,
+            this.modelVars[cellPrefix + 'bias'] as tf.Tensor1D,
+            data, c, h
+          ));
+    }
+
+    this.resetState();
+
+    this.initialized = true;
+  }
+
+  /**
+   * Evaluates Piano Genie to produce piano key logits and update state.
+   *
+   * @param button Button number [0, 8)
+   * @param lastOutput Last sampled piano key [0, 88). Pass -1 on first call.
+   * @param deltaTime Amount of time elapsed in seconds since last button press.
+   */
+  evaluate(button: number, lastOutput: number, deltaTime: number) {
```

Comment:
if the last output is needed for the model to make a next prediction, i think that the model, not the user, should keep track of it.